chain/network/src/peer_manager/peer_manager_actor.rs

```diff
@@ -1063,90 +1120,84 @@ impl PeerManagerActor {
+            <= self.config.ideal_connections_hi as usize
         {
-            for (peer, active) in self.connected_peers.iter() {
-                if active.peer_type == PeerType::Outbound {
-                    safe_set.insert(peer);
-                }
-            }
+            return;
         }
+        // Add whitelisted nodes to the safe set.
+        safe_set.extend(whitelisted_peers.into_iter().cloned());
 
-        if self.config.archive
-            && (self.connected_peers.values())
-                .filter(|connected_peer| connected_peer.full_peer_info.chain_info.archival)
-                .count()
-                <= self.config.archival_peer_connections_lower_bound as usize
+        // If there is not enough outbound peers, add them to the safe set.
+        let outbound_peers = filter_peers(&|p| p.peer_type == PeerType::Outbound);
+        if outbound_peers.len() + self.outgoing_peers.len()
+            <= self.config.minimum_outbound_peers as usize
         {
-            for (peer, active) in self.connected_peers.iter() {
-                if active.full_peer_info.chain_info.archival {
-                    safe_set.insert(peer);
-                }
-            }
+            safe_set.extend(outbound_peers.into_iter().cloned());
         }
 
-        // Find all recent connections
-        let mut recent_connections = (self.connected_peers.iter())
-            .filter_map(|(peer_id, active)| {
-                if active.last_time_received_message.elapsed() < self.config.peer_recent_time_window
-                {
-                    Some((peer_id.clone(), active.connection_established_time))
-                } else {
-                    None
-                }
-            })
-            .collect::<Vec<_>>();
-
-        // Sort by established time
-        recent_connections.sort_by(|(_, established_time_a), (_, established_time_b)| {
-            established_time_a.cmp(established_time_b)
-        });
+        // If there is not enough archival peers, add them to the safe set.
+        if self.config.archive {
+            let archival_peers = filter_peers(&|p| p.full_peer_info.chain_info.archival);
+            if archival_peers.len() <= self.config.archival_peer_connections_lower_bound as usize {
+                safe_set.extend(archival_peers.into_iter().cloned());
+            }
+        }
 
-        // Take remaining peers
-        for (peer_id, _) in recent_connections
+        // Find all recently active peers.
+        let mut active_peers: Vec<_> = self
+            .connected_peers
             .iter()
-            .take((self.config.safe_set_size as usize).saturating_sub(safe_set.len()))
-        {
-            safe_set.insert(peer_id);
+            .filter(|(_, p)| {
+                p.last_time_received_message.elapsed() < self.config.peer_recent_time_window
+            })
+            .collect();
+        // Sort by established time.
+        active_peers.sort_by_key(|(_, p)| p.connection_established_time);
```

Comment:
By the way, if you’d filter peers which are already in safe set (`!safe_set.contains(id) && p.last_time_received_message.elapsed() < self.config.peer_recent_time_window`) you could use [quick select](https://doc.rust-lang.org/std/primitive.slice.html#method.select_nth_unstable_by_key) to further optimise the sorting.  Something like:

```rust
        let needed = (self.config.safe_set_size as usize).saturating_sub(safe_set.len());
        if needed > 0 {
            let mut active_peers: Vec<_> = self
                .connected_peers
                .iter()
                .filter(|(id, info)| {
                    info.last_time_received_message.elapsed() < self.config.peer_recent_time_window &&
                        !safe_set.contains(id)
                })
                .collect();
            let active_peers = if needed <= active_peers.len() {
                active_peers.as_slice()
            } else {
                active_peers.select_nth_unstable_by_key(needed - 1, |(_, p)| p.connection_established_time);
                &active_peers[..needed]
            };
            safe_set.extend(active_peers.into_iter().map(|(id, _)| id.clone()));
        }
```

Reply:
in general, sure. But with ~40 elements array I doubt that it makes any difference ;)